pkg/cmd/repo/deploy-key/add/add.go

```diff
@@ -0,0 +1,112 @@
+package add
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+
+	"github.com/cli/cli/v2/internal/config"
+	"github.com/cli/cli/v2/internal/ghrepo"
+	"github.com/cli/cli/v2/pkg/cmdutil"
+	"github.com/cli/cli/v2/pkg/iostreams"
+	"github.com/spf13/cobra"
+)
+
+type AddOptions struct {
+	IO         *iostreams.IOStreams
+	Config     func() (config.Config, error)
+	HTTPClient func() (*http.Client, error)
+	BaseRepo   func() (ghrepo.Interface, error)
+
+	KeyFile   string
+	Title     string
+	ReadWrite bool
+}
+
+func NewCmdAdd(f *cmdutil.Factory, runF func(*AddOptions) error) *cobra.Command {
+	opts := &AddOptions{
+		HTTPClient: f.HttpClient,
+		Config:     f.Config,
+		IO:         f.IOStreams,
+		BaseRepo:   f.BaseRepo,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "add [<key-file>] [--read-write]",
+		Short: "Add a deploy key to your GitHub repository",
+		Args:  cobra.MaximumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				if opts.IO.IsStdoutTTY() && opts.IO.IsStdinTTY() {
+					return &cmdutil.FlagError{Err: errors.New("public key file missing")}
+				}
+				opts.KeyFile = "-"
+			} else {
+				opts.KeyFile = args[0]
+			}
+
+			if runF != nil {
+				return runF(opts)
+			}
+			return addRun(opts)
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.Title, "title", "t", "", "Title for the new key")
+	cmd.Flags().BoolVarP(&opts.ReadWrite, "read-write", "w", false, "Allow write access")
+	return cmd
+}
+
+func addRun(opts *AddOptions) error {
+	httpClient, err := opts.HTTPClient()
+	if err != nil {
+		return err
+	}
+
+	var keyReader io.Reader
+	if opts.KeyFile == "-" {
+		keyReader = opts.IO.In
+		defer opts.IO.In.Close()
+	} else {
+		f, err := os.Open(opts.KeyFile)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		keyReader = f
+	}
+
+	cfg, err := opts.Config()
+	if err != nil {
+		return err
+	}
+
+	hostname, err := cfg.DefaultHost()
+	if err != nil {
+		return err
+	}
+
+	repo, err := opts.BaseRepo()
+	if err != nil {
+		return err
+	}
+
+	err = uploadDeployKey(httpClient, hostname, keyReader, opts, repo)
+	if err != nil {
+		if errors.Is(err, scopesError) {
+			cs := opts.IO.ColorScheme()
+			fmt.Fprint(opts.IO.ErrOut, "Error: insufficient OAuth scopes to add deploy keys\n")
+			fmt.Fprintf(opts.IO.ErrOut, "Run the following to grant scopes: %s\n", cs.Bold("gh auth refresh -s write:public_key"))
+			return cmdutil.SilentError
+		}
+		return err
+	}
+
+	if opts.IO.IsStdoutTTY() {
+		cs := opts.IO.ColorScheme()
+		fmt.Fprintf(opts.IO.Out, "%s Public key added to your repository\n", cs.SuccessIcon())
```

Comment:
Instead of "your repository", let's print the actual repository in `owner/repo` format